### PR TITLE
Implement profiler

### PR DESCRIPTION
**Pull request by ZehMatt**

### CODE REVIEW

**Review by tupaschoal — COMMENTED** [review]
Out of curiosity, did you consider using something like https://github.com/bombomby/optick ?

**Comment by ZehMatt** [comment]
> Out of curiosity, did you consider using something like https://github.com/bombomby/optick ?

I did not, I wanted this to be as lightweight as possible and in my experience most thirdparty libraries barely fall into this category.

File: src/openrct2/entity/EntityRegistry.cpp [hunk 1]
```diff
@@ -19,6 +19,7 @@
 #include "../entity/Staff.h"
 #include "../interface/Viewport.h"
 #include "../peep/RideUseSystem.h"
+#include "../profiling/Profiling.h"
```
Comment (Gymnasiast) [inline src/openrct2/entity/EntityRegistry.cpp:22]: You are including the header here, but this file does not seem to have any profiled functions.

Comment (ZehMatt) [reply src/openrct2/entity/EntityRegistry.cpp:22]: You are right, this probably slipped on the rebase.

File: src/openrct2/interface/InteractiveConsole.cpp [hunk 2]
```diff
@@ -1829,6 +1869,10 @@ static constexpr const console_command console_command_table[] = {
       "replay_normalise <input file> <output file>" },
     { "mp_desync", cc_mp_desync, "Forces a multiplayer desync",
       "cc_mp_desync [desync_type, 0 = Random t-shirt color on random guest, 1 = Remove random guest ]" },
+    { "profiler_reset", cc_profiler_reset, "Resets the profiler data.", "profiler_reset" },
+    { "profiler_start", cc_profiler_start, "Starts the profiler.", "profiler_start" },
+    { "profiler_stop", cc_profiler_stop, "Stops the profiler.", "profiler_stop" },
+    { "profiler_exportcsv", cc_profiler_exportcsv, "Exports the current profiler data.", "profiler_exportcsv" },
```
Comment (janisozaur) [inline src/openrct2/interface/InteractiveConsole.cpp:1875]:
```suggestion
    { "profiler_exportcsv", cc_profiler_exportcsv, "Exports the current profiler data.", "profiler_exportcsv <output file>" },
```

File: src/openrct2/interface/InteractiveConsole.cpp [hunk 3]
```diff
@@ -1829,6 +1869,10 @@ static constexpr const console_command console_command_table[] = {
       "replay_normalise <input file> <output file>" },
     { "mp_desync", cc_mp_desync, "Forces a multiplayer desync",
       "cc_mp_desync [desync_type, 0 = Random t-shirt color on random guest, 1 = Remove random guest ]" },
+    { "profiler_reset", cc_profiler_reset, "Resets the profiler data.", "profiler_reset" },
+    { "profiler_start", cc_profiler_start, "Starts the profiler.", "profiler_start" },
+    { "profiler_stop", cc_profiler_stop, "Stops the profiler.", "profiler_stop" },
```
Comment (janisozaur) [inline src/openrct2/interface/InteractiveConsole.cpp:1874]: this could take optional param to output file and write it once stopped

**Review by janisozaur — COMMENTED** [review]
I think it mostly looks correct other than those minor UX issues. I would like to test it tomorrow.

File: src/openrct2/interface/InteractiveConsole.cpp [hunk 4]
```diff
@@ -1725,6 +1726,45 @@ static int32_t cc_add_news_item([[maybe_unused]] InteractiveConsole& console, [[
...
+        return 1;
+    }
+
+    if (!OpenRCT2::Profiling::ExportCSV(argv[0]))
+    {
+        console.WriteLineError("Missing argument: <file path>");
+        return 1;
+    }
+
+    console.WriteLine("Wrote file!");
```
Comment (janisozaur) [inline src/openrct2/interface/InteractiveConsole.cpp:1764]: Can it report absolute filename?

**Comment by janisozaur** [comment]
I have no other comments than the ones left already

**Review by janisozaur — APPROVED** [review]

**Pull request merged**